Add support for aria placeholder

feerrenrut opened the pull request:
This PR intends to introduce [aria-placeholder](https://w3c.github.io/aria/#aria-placeholder) support to NVDA. 

From the spec it seems unclear whether or the placeholder attribute will be removed (by the browser or the web author, [see issue](https://github.com/w3c/aria/issues/541) ) when the user enters the value. Because of this, it is implemented in NVDA.

Here is an overview of the design. The wiki has some other relevant [notes on implementing aria-current](https://github.com/nvaccess/nvda/wiki/Notes:-Adding-support-for-aria-attributes-to-browsers), that gives some more background.
```
class NVDAObject (source/NVDAObjects/__init__.py)
|	def _get__isTextEmpty(self): has default impl that does not consider space or newline
|	def _get_placeholder(self): returns None
|
+-> class EdgeNode > UIA > window.Window > NVDAObject (source/NVDAObjects/UIA/edge.py)
|			def _get__isTextEmpty(self)
|			def _get_placeholder(self)
|
+-> class Ia2Web > IAccessible > window.Window > NVDAObject (source/NVDAObjects/IAccessible/ia2Web.py)
|			def _get_placeholder(self)
|
+-> class MSHTML > IAccessible > window.Window > NVDAObject (source/NVDAObjects/IAccessible/MSHTML.py)
			def _get_placeholder(self):

class VirtualBufferTextInfo (source/virtualBuffers/__init__.py)
|	def _getPlaceholderAttribute(self, attrs, placeholderAttrsKey)
|
+-> class Gecko_ia2_TextInfo -> VirtualBufferTextInfo (source/virtualBuffers/gecko_ia2.py)
|			def _normalizeControlField(self,attrs) [calls] self._getPlaceholderAttribute
|
+-> class MSHTMLTextInfo -> VirtualBufferTextInfo
			def _normalizeControlField(self,attrs) [calls] self._getPlaceholderAttribute

class EdgeTextInfo > UIATextInfo > textInfos.TextInfo (source/NVDAObjects/UIA/edge.py)
			def _get_placeholder(self)
			def _getControlFieldForObject [calls] def _get_placeholder(self)
```

In terms of UX, the placeholder value is not prefixed by any other text when reported in either speech or braille. It is only reported when the field does not have a value. 

For now we will not be supporting 'role=searchbox` despite this being given as an example on the Aria spec. This decision was made because the browser implementations aren't really ready yet

For the testing performed please see [this comment](https://github.com/nvaccess/nvda/pull/7055#issuecomment-303925221)

feerrenrut committed:
Add support for aria-placeholder

Fixes #7004

Handles values with backslash escaped semicolons and backslash escaped
backslashes.

feerrenrut committed:
Implement browse mode for chrome

Dont report placeholder when a value is present.

feerrenrut committed:
Tidy up code

feerrenrut commented:
Using [aria-placeholder-test.zip](https://github.com/nvaccess/nvda/files/904726/aria-placeholder-test.zip) to test.

Chrome
- when there is a space in the `input type="text"` example, in browse mode placeholder is still reported. This is not the case for the `div contenteditable` example

Edge
- in focus mode
 - in the `div contenteditiable` example
  - placeholder is reported as "grouping placeholder: blah"
  - the value is not reported / selected
 - in the `input type="text"` example
 - placeholder is reported as "placeholder: blah"
  - the value is reported / selected
- in browse mode
 - the value is reported in both examples
 - the placeholder value is not reported in either example

Firefox
- browse mode
 - The placeholder value is not reported in the `div contenteditable` examples
- Focus mode behaves as expected

IE
- Focus mode
 - content editable examples do not report the value, do report placeholder
 - `input type="text"` examples work as expected.
- Browse mode
 - placeholder not reported.

feerrenrut commented:
Using Edge with nvda-next (EG without any of the modification in this branch) while in focus mode the values of `div contenteditable` fields are not reported and nothing is selected when tabing to the fields. However "grouping" is reported. For `input type="text"` fields the value is selected and reported in focus mode.

jcsteh commented on source/virtualBuffers/gecko_ia2.py:
```diff
@@ -23,9 +23,24 @@
 		ariaCurrent = attrs.get("IAccessible2::attribute_current")
 		if ariaCurrent != None:
 			attrs['current']= ariaCurrent
+		ariaPlaceholder = attrs.get("IAccessible2::attribute_placeholder")
```
Rename to just `placeholder`.

jcsteh commented on source/virtualBuffers/gecko_ia2.py:
```diff
@@ -23,9 +23,24 @@
 			attrs['current']= ariaCurrent
+		ariaPlaceholder = attrs.get("IAccessible2::attribute_placeholder")
+		# For efficiency, only check if it is valid to include placeholder when we have a placeholder value to include.
+		if ariaPlaceholder != None:
```
Change this to either `if placeholder` or (if you're worried about the difference between None and empty string) `if placeholder is not None`.

jcsteh commented on source/virtualBuffers/gecko_ia2.py:
```diff
@@ -23,9 +23,24 @@
+				pass
+			else:
+				# only allow placeholder attribute if the field value is empty
+				controlFieldText = self.obj.makeTextInfo(textInfos.offsets.Offsets(start, end)).text
```
Since you have the offsets, you can avoid potentially fetching a huge amount of text by checking if the offsets span only one character. Even an empty field will render a space in browse mode, so you'll still need to check for `" "`.

jcsteh reviewed (commented):
I realise this is still work in progress, but here's some initial feedback. Overall, this looks good.

jcsteh commented:
Also, a reminder to check empty value for braille.

feerrenrut committed:
review actions for #7055

Fix Chrome browse mode with div contenteditable

In this case, `controlFieldText` matches the label for the field.
Meaning that we were not reporting the placeholder since there is a "value"
placeholder is now reported when the value.

feerrenrut committed:
Remove workaround to report placeholder when value==label

This has been split out into another bug, see #7153

feerrenrut commented:
## Re-testing aria-placeholder.

### Testing two (empty value) controls:
1. `<input type="text" role="searchbox" aria-label="myLabel1" aria-placeholder="my placeholder" value="">`
2. `<div contenteditable role="searchbox" aria-label="myLabel2" aria-placeholder="my placeholder" stlye="height:10px;width:30px"></div>`

Its expected that the following is reported:
Focus mode:
1. "myLabel1  edit  my placeholder blank"
2. "myLabel2  edit  multi line  my placeholder  blank"

Browse mode:
1. "edit  clickable  my placeholder"
2. "edit  multi line  clickable  my placeholder"

### Results:
Chrome - Focus mode:
1. "myLabel1  edit  my placeholder  blank"
2. "myLabel2  edit  multi line  my placeholder  blank"

Chrome - Browse mode:
1. "edit  clickable  my placeholder"
2. "edit  multi line  clickable  my placeholder"

Firefox - Focus mode:
1. "myLabel1  edit  has auto complete my placeholder  blank"
2. "myLabel2  edit  my placeholder  blank"

Firefox - Browse mode:
1. "edit  has auto complete  my placeholder"
2. Cant get to the field with browse mode.

Edge - Focus mode:
1. "myLabel1  edit  my placeholder  blank"
2. "blank"

Edge - Browse mode:
1. "blank  edit"
2. "blank"

Then the tab "crashes". The page is replaced by an edge error "this page is having a problem loading."
This page seems to work fine without nvda loaded.

feerrenrut committed:
Fix some issues in Edge and IE

feerrenrut commented:
My last testing comment has been updated with new results. Summary:
There is still no reporting of placeholder value for `div contentEditable` fields from browse mode for Edge and IE. All other cases are working as expected.

For edge this seems to be because the `div contentEditable` field has `isEditable==False` in `speakObject`

feerrenrut committed:
Extraxt empty checking logic to NVDAObject

In focus mode we now check if the content of the field is empty via the
NVDAObject _get__isTextEmpty method.

In browse mode, extracted and shared code to get the placeholder
attribute if the content of the field is empty. This is used by IE,
Firefox, and Chrome.

Fixed up some braille and speech cases.

feerrenrut commented:
I have just pushed some further changes to this branch, below is the latest set of test results using the following test file:
[aria-placeholder-test.html.txt](https://github.com/nvaccess/nvda/files/1028023/aria-placeholder-test.html.txt)

Tests:

1. div contenteditable role=textbox with text already entered:
2. div contenteditable role=textbox with no text entered:
3. div contenteditable with text already entered:
4. div contenteditable with no text entered:
5. input type=text role=textbox with text already entered:
6. input type=text role=textbox with no text entered:
7. input type=textarea with text already entered:
8. input type=textarea with no text entered:

Browsers:

- Chrome Canary
- Firefox
- Edge
- IE

The "Note" column indicates where there is some suspicious looking difference.
All these note markings seem to be pointing to "space" being reported.
Unfortunately when testing I didn't synchronise my config between the two builds.
On my "next" build I had "Line indentation reporting" set to speech.

### 1. div contenteditable role=textbox with text already entered:

|Note|Browser|NVDA Mode|Ouput Type|Branch|Output
|----|-------|---------|----------|------|------
| |Chrome |Focus    |Speech    |Next  |div contenteditable role=textbox with text already entered:  edit  multi line  03-14-1879
| |       |         |          |PR    |div contenteditable role=textbox with text already entered:  edit  multi line  03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder document ', u'div contenteditable role=textbox with text already entered: multi line edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder document ', u'div contenteditable role=textbox with text already entered: multi line edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  multi line  03-14-1879
| |       |         |          |PR    |edit  multi line  03-14-1879
| |       |         |Braille   |Next  |[u'multi line edt 03-14-1879 edt end']
| |       |         |          |PR    |[u'multi line edt 03-14-1879 edt end']
| |FireFox|Focus    |Speech    |Next  |div contenteditable role=textbox with text already entered:  edit  03-14-1879
| |       |         |          |PR    |div contenteditable role=textbox with text already entered:  edit  03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with text already entered: edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with text already entered: edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  03-14-1879
| |       |         |          |PR    |edit  03-14-1879
| |       |         |Braille   |Next  |[u'edt 03-14-1879 ']
| |       |         |          |PR    |[u'edt 03-14-1879 ']
| |Edge   |Focus    |Speech    |Next  |div contenteditable role=textbox with text already entered:  edit  edit  03-14-1879
| |       |         |          |PR    |div contenteditable role=textbox with text already entered:  edit  03-14-1879
| |       |         |Braille   |Next  |[u'edt 03-14-1879 ']
|x|       |         |          |PR    |[u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  03-14-1879
| |       |         |          |PR    |edit  03-14-1879
| |       |         |Braille   |Next  |[u'edt 03-14-1879 ']
| |       |         |          |PR    |[u'edt 03-14-1879 ']
| |IE     |Focus    |Speech    |Next  |div contenteditable role=textbox with text already entered:  edit
| |       |         |          |PR    |div contenteditable role=textbox with text already entered:  edit
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with text already entered: edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with text already entered: edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  03-14-1879
| |       |         |          |PR    |edit  03-14-1879
| |       |         |Braille   |Next  |[u'edt 03-14-1879 ']
| |       |         |          |PR    |[u'edt 03-14-1879 ']

### 2. div contenteditable role=textbox with no text entered:

|Note|Browser|NVDA Mode|Ouput Type|Branch|Output
|----|-------|---------|----------|------|------
| |Chrome |Focus    |Speech    |Next  |div contenteditable role=textbox with no text entered:  edit  multi line  blank
| |       |         |          |PR    |div contenteditable role=textbox with no text entered:  edit  multi line  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder document ', u'div contenteditable role=textbox with no text entered: multi line edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder document ', u'div contenteditable role=textbox with no text entered: multi line edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  multi line  space
|x|       |         |          |PR    |edit  multi line  my placeholder
| |       |         |Braille   |Next  |[u'multi line edt   edt end']
| |       |         |          |PR    |[u'multi line edt my placeholder   edt end']
| |FireFox|Focus    |Speech    |Next  |div contenteditable role=textbox with no text entered:  edit  blank
| |       |         |          |PR    |div contenteditable role=textbox with no text entered:  edit  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with no text entered: edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with no text entered: edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  space
|x|       |         |          |PR    |edit  my placeholder
| |       |         |Braille   |Next  |[u'edt   ']
| |       |         |          |PR    |[u'edt my placeholder   ']
| |Edge   |Focus    |Speech    |Next  |div contenteditable role=textbox with no text entered:  edit  blank
| |       |         |          |PR    |div contenteditable role=textbox with no text entered:  edit  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with no text entered: edt ', u' ']
| |       |         |          |PR    |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with no text entered: edt my placeholder ', u' ']
| |       |Browse   |Speech    |Next  |edit
| |       |         |          |PR    |edit  my placeholder
| |       |         |Braille   |Next  |[u'edt  ']
| |       |         |          |PR    |[u'edt my placeholder  ']
| |IE     |Focus    |Speech    |Next  |div contenteditable role=textbox with no text entered:  edit
| |       |         |          |PR    |div contenteditable role=textbox with no text entered:  edit
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with no text entered: edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'div contenteditable role=textbox with no text entered: edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  space
|x|       |         |          |PR    |edit  my placeholder
| |       |         |Braille   |Next  |[u'edt   ']
| |       |         |          |PR    |[u'edt my placeholder   ']

### 3. div contenteditable with text already entered:

|Note|Browser|NVDA Mode|Ouput Type|Branch|Output
|----|-------|---------|----------|------|------
| |Chrome |Focus    |Speech    |Next  |div contenteditable with text already entered:  section  editable  03-14-1879
| |       |         |          |PR    |div contenteditable with text already entered:  section  editable  03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder document ', u'div contenteditable with text already entered: editable section ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder document ', u'div contenteditable with text already entered: editable section ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |section  editable  03-14-1879
| |       |         |          |PR    |section  editable  03-14-1879
| |       |         |Braille   |Next  |[u'editable section 03-14-1879 section end']
| |       |         |          |PR    |[u'editable section 03-14-1879 section end']
| |FireFox|Focus    |Speech    |Next  |div contenteditable with text already entered:  section  editable  03-14-1879
| |       |         |          |PR    |div contenteditable with text already entered:  section  editable  03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'div contenteditable with text already entered: editable section ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'div contenteditable with text already entered: editable section ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |section  editable  03-14-1879
| |       |         |          |PR    |section  editable  03-14-1879
| |       |         |Braille   |Next  |[u'editable section 03-14-1879 section end']
| |       |         |          |PR    |[u'editable section 03-14-1879 section end']
| |Edge   |Focus    |Speech    |Next  |div contenteditable with text already entered:  grouping
| |       |         |          |PR    |div contenteditable with text already entered:  grouping
| |       |         |Braille   |Next  |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'div contenteditable with text already entered: grouping']
| |       |         |          |PR    |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'div contenteditable with text already entered: grouping']
| |       |Browse   |Speech    |Next  |03-14-1879
| |       |         |          |PR    |03-14-1879
| |       |         |Braille   |Next  |[u'03-14-1879 ']
| |       |         |          |PR    |[u'03-14-1879 ']
| |IE     |Focus    |Speech    |Next  |div contenteditable with text already entered:  grouping  editable
| |       |         |          |PR    |div contenteditable with text already entered:  grouping  editable
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'div contenteditable with text already entered: editable grouping ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'div contenteditable with text already entered: editable grouping ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |section  editable  03-14-1879
| |       |         |          |PR    |section  editable  03-14-1879
| |       |         |Braille   |Next  |[u'editable section 03-14-1879 section end']
| |       |         |          |PR    |[u'editable section 03-14-1879 section end']

### 4. div contenteditable with no text entered:

|Note|Browser|NVDA Mode|Ouput Type|Branch|Output
|----|-------|---------|----------|------|------
| |Chrome |Focus    |Speech    |Next  |div contenteditable with no text entered:  section  editable  blank
| |       |         |          |PR    |div contenteditable with no text entered:  section  editable  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder document ', u'div contenteditable with no text entered: editable section ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder document ', u'div contenteditable with no text entered: editable section my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |section  editable  div contenteditable with no text entered:
| |       |         |          |PR    |section  editable  div contenteditable with no text entered:
| |       |         |Braille   |Next  |[u'editable section div contenteditable with no text entered: section end']
| |       |         |          |PR    |[u'editable section div contenteditable with no text entered: section end']
| |FireFox|Focus    |Speech    |Next  |div contenteditable with no text entered:  section  editable  blank
| |       |         |          |PR    |div contenteditable with no text entered:  section  editable  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'div contenteditable with no text entered: editable section ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'div contenteditable with no text entered: editable section my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |section  editable  div contenteditable with no text entered:
| |       |         |          |PR    |section  editable  div contenteditable with no text entered:
| |       |         |Braille   |Next  |[u'editable section div contenteditable with no text entered: section end']
| |       |         |          |PR    |[u'editable section div contenteditable with no text entered: section end']
| |Edge   |Focus    |Speech    |Next  |div contenteditable with no text entered:  grouping
| |       |         |          |PR    |div contenteditable with no text entered:  grouping
| |       |         |Braille   |Next  |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'div contenteditable with no text entered: grouping']
| |       |         |          |PR    |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'div contenteditable with no text entered: grouping my placeholder']
| |       |Browse   |Speech    |Next  |blank
| |       |         |          |PR    |blank
| |       |         |Braille   |Next  |[u' ']
| |       |         |          |PR    |[u' ']
| |IE     |Focus    |Speech    |Next  |div contenteditable with no text entered:  grouping  editable
| |       |         |          |PR    |div contenteditable with no text entered:  grouping  editable
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'div contenteditable with no text entered: editable grouping ', u'div contenteditable with no text entered: ']
| |       |         |          |PR    |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'div contenteditable with no text entered: editable grouping ', u'div contenteditable with no text entered: ']
| |       |Browse   |Speech    |Next  |section  editable  space
|x|       |         |          |PR    |section  editable  my placeholder
| |       |         |Braille   |Next  |[u'editable section   section end']
| |       |         |          |PR    |[u'editable section my placeholder   section end']

### 5. input type=text role=textbox with text already entered:

|Note|Browser|NVDA Mode|Ouput Type|Branch|Output
|----|-------|---------|----------|------|------
| |Chrome |Focus    |Speech    |Next  |input type=text role=textbox with text already entered:  edit  selected 03-14-1879
| |       |         |          |PR    |input type=text role=textbox with text already entered:  edit  selected 03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder document ', u'input type=text role=textbox with text already entered: edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder document ', u'input type=text role=textbox with text already entered: edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  03-14-1879
| |       |         |          |PR    |edit  03-14-1879
| |       |         |Braille   |Next  |[u'edt 03-14-1879 ']
| |       |         |          |PR    |[u'edt 03-14-1879 ']
| |FireFox|Focus    |Speech    |Next  |input type=text role=textbox with text already entered:  edit  has auto complete  selected 03-14-1879
| |       |         |          |PR    |input type=text role=textbox with text already entered:  edit  has auto complete  selected 03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'input type=text role=textbox with text already entered: ... edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'input type=text role=textbox with text already entered: ... edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  has auto complete  03-14-1879
| |       |         |          |PR    |edit  has auto complete  03-14-1879
| |       |         |Braille   |Next  |[u'... edt 03-14-1879 ']
| |       |         |          |PR    |[u'... edt 03-14-1879 ']
| |Edge   |Focus    |Speech    |Next  |input type=text role=textbox with text already entered:  edit  selected 03-14-1879
| |       |         |          |PR    |input type=text role=textbox with text already entered:  edit  selected 03-14-1879
| |       |         |Braille   |Next  |[u'03-14-1879 ']
| |       |         |          |PR    |[u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |blank
| |       |         |          |PR    |blank
| |       |         |Braille   |Next  |[' ']
| |       |         |          |PR    |[' ']
| |IE     |Focus    |Speech    |Next  |input type=text role=textbox with text already entered:  edit  selected 03-14-1879
| |       |         |          |PR    |input type=text role=textbox with text already entered:  edit  selected 03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'input type=text role=textbox with text already entered: edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'input type=text role=textbox with text already entered: edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  03-14-1879
| |       |         |          |PR    |edit  03-14-1879
| |       |         |Braille   |Next  |[u'edt 03-14-1879 ']
| |       |         |          |PR    |[u'edt 03-14-1879 ']

### 6. input type=text role=textbox with no text entered:

|Note|Browser|NVDA Mode|Ouput Type|Branch|Output
|----|-------|---------|----------|------|------
| |Chrome |Focus    |Speech    |Next  |input type=text role=textbox with no text entered:  edit  blank
| |       |         |          |PR    |input type=text role=textbox with no text entered:  edit  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder document ', u'input type=text role=textbox with no text entered: edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder document ', u'input type=text role=textbox with no text entered: edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  space
|x|       |         |          |PR    |edit  my placeholder
| |       |         |Braille   |Next  |[u'edt   ']
| |       |         |          |PR    |[u'edt my placeholder   ']
| |FireFox|Focus    |Speech    |Next  |input type=text role=textbox with no text entered:  edit  has auto complete  blank
| |       |         |          |PR    |input type=text role=textbox with no text entered:  edit  has auto complete  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'input type=text role=textbox with no text entered: ... edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'input type=text role=textbox with no text entered: ... edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  has auto complete  space
|x|       |         |          |PR    |edit  has auto complete  my placeholder
| |       |         |Braille   |Next  |[u'... edt   ']
| |       |         |          |PR    |[u'... edt my placeholder   ']
| |Edge   |Focus    |Speech    |Next  |input type=text role=textbox with no text entered:  edit  blank
| |       |         |          |PR    |input type=text role=textbox with no text entered:  edit  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'input type=text role=textbox with no text entered: edt ', u' ']
| |       |         |          |PR    |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'input type=text role=textbox with no text entered: edt my placeholder ', u' ']
| |       |Browse   |Speech    |Next  |blank
| |       |         |          |PR    |blank
| |       |         |Braille   |Next  |[' ']
| |       |         |          |PR    |[' ']
| |IE     |Focus    |Speech    |Next  |input type=text role=textbox with no text entered:  edit  blank
| |       |         |          |PR    |input type=text role=textbox with no text entered:  edit  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'input type=text role=textbox with no text entered: edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'input type=text role=textbox with no text entered: edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  space
|x|       |         |          |PR    |edit  my placeholder
| |       |         |Braille   |Next  |[u'edt   ']
| |       |         |          |PR    |[u'edt my placeholder   ']

### 7. input type=textarea with text already entered:

|Note|Browser|NVDA Mode|Ouput Type|Branch|Output
|----|-------|---------|----------|------|------
| |Chrome |Focus    |Speech    |Next  |input type=textarea with text already entered:  edit  selected 03-14-1879
| |       |         |          |PR    |input type=textarea with text already entered:  edit  selected 03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder document ', u'input type=textarea with text already entered: edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder document ', u'input type=textarea with text already entered: edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  03-14-1879
| |       |         |          |PR    |edit  03-14-1879
| |       |         |Braille   |Next  |[u'edt 03-14-1879 ']
| |       |         |          |PR    |[u'edt 03-14-1879 ']
| |FireFox|Focus    |Speech    |Next  |input type=textarea with text already entered:  edit  has auto complete  selected 03-14-1879
| |       |         |          |PR    |input type=textarea with text already entered:  edit  has auto complete  selected 03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'input type=textarea with text already entered: ... edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'input type=textarea with text already entered: ... edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  has auto complete  03-14-1879
| |       |         |          |PR    |edit  has auto complete  03-14-1879
| |       |         |Braille   |Next  |[u'... edt 03-14-1879 ']
| |       |         |          |PR    |[u'... edt 03-14-1879 ']
| |Edge   |Focus    |Speech    |Next  |input type=textarea with text already entered:  edit  selected 03-14-1879
| |       |         |          |PR    |input type=textarea with text already entered:  edit  selected 03-14-1879
| |       |         |Braille   |Next  |[u'03-14-1879 ']
| |       |         |          |PR    |[u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |blank
| |       |         |          |PR    |blank
| |       |         |Braille   |Next  |[' ']
| |       |         |          |PR    |[' ']
| |IE     |Focus    |Speech    |Next  |input type=textarea with text already entered:  edit  selected 03-14-1879
| |       |         |          |PR    |input type=textarea with text already entered:  edit  selected 03-14-1879
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'input type=textarea with text already entered: edt ', u'03-14-1879 ']
| |       |         |          |PR    |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'input type=textarea with text already entered: edt ', u'03-14-1879 ']
| |       |Browse   |Speech    |Next  |edit  03-14-1879
| |       |         |          |PR    |edit  03-14-1879
| |       |         |Braille   |Next  |[u'edt 03-14-1879 ']
| |       |         |          |PR    |[u'edt 03-14-1879 ']

### 8. input type=textarea with no text entered:

|Note|Browser|NVDA Mode|Ouput Type|Branch|Output
|----|-------|---------|----------|------|------
| |Chrome |Focus    |Speech    |Next  |input type=textarea with no text entered:  edit  blank
| |       |         |          |PR    |input type=textarea with no text entered:  edit  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder document ', u'input type=textarea with no text entered: edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder document ', u'input type=textarea with no text entered: edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  space
|x|       |         |          |PR    |edit  my placeholder
| |       |         |Braille   |Next  |[u'edt   ']
| |       |         |          |PR    |[u'edt my placeholder   ']
| |FireFox|Focus    |Speech    |Next  |input type=textarea with no text entered:  edit  has auto complete  blank
| |       |         |          |PR    |input type=textarea with no text entered:  edit  has auto complete  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'input type=textarea with no text entered: ... edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder - Mozilla Firefox ', u'Test aria-placeholder document ', u'input type=textarea with no text entered: ... edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  has auto complete  space
|x|       |         |          |PR    |edit  has auto complete  my placeholder
| |       |         |Braille   |Next  |[u'... edt   ']
| |       |         |          |PR    |[u'... edt my placeholder   ']
| |Edge   |Focus    |Speech    |Next  |input type=textarea with no text entered:  edit  blank
| |       |         |          |PR    |input type=textarea with no text entered:  edit  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'input type=textarea with no text entered: edt ', u' ']
| |       |         |          |PR    |[u'Test aria-placeholder \u200e- Microsoft Edge ', u'Microsoft Edge window ', u'Test aria-placeholder document ', u'input type=textarea with no text entered: edt my placeholder ', u' ']
| |       |Browse   |Speech    |Next  |blank
| |       |         |          |PR    |blank
| |       |         |Braille   |Next  |[' ']
| |       |         |          |PR    |[' ']
| |IE     |Focus    |Speech    |Next  |input type=textarea with no text entered:  edit  blank
| |       |         |          |PR    |input type=textarea with no text entered:  edit  my placeholder  blank
| |       |         |Braille   |Next  |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'input type=textarea with no text entered: edt ', ' ']
| |       |         |          |PR    |[u'Test aria-placeholder - Internet Explorer ', u'Test aria-placeholder document ', u'input type=textarea with no text entered: edt my placeholder ', ' ']
| |       |Browse   |Speech    |Next  |edit  space
|x|       |         |          |PR    |edit  my placeholder
| |       |         |Braille   |Next  |[u'edt   ']
| |       |         |          |PR    |[u'edt my placeholder   ']

feerrenrut commented:
Some previous discussion that should be documented here. For now we will not be supporting 'role=searchbox` despite this being given as an example on the Aria spec. This decision was made because the browser implementations aren't really ready yet.

feerrenrut commented on source/virtualBuffers/__init__.py:
```diff
@@ -218,6 +218,37 @@ def _getTextRange(self,start,end):
+			# could hold space or \n)
+			if valueLen < 2:
+				controlFieldText = self.obj.makeTextInfo(textInfos.offsets.Offsets(start, end)).text
+				if not controlFieldText or controlFieldText == ' ' or controlFieldText == '\n':
```
Not actually 100% sure that we need to have exceptions for space and newline here.

jcsteh replied:
If it's truly a blank field, the only thing we'll ever insert is a space. If it contains a new line, that means the field really does contain a new line. So you should just check for space.

feerrenrut replied:
Yep, did this and tested against Chrome and Firefox in browse mode.

feerrenrut commented:
The missing 'edt' in Edge focus mode in the first test, seems to be a separate issue, and I think has been independently fixed in next. I tested the commit prior to any of my placeholder changes, which had the same behaviour of omitting 'edt'.

jcsteh commented on source/speech.py:
```diff
@@ -308,16 +308,29 @@ def speakObjectProperties(obj,reason=controlTypes.REASON_QUERY,index=None,**allo
 	if text:
 		speakText(text,index=index)
 
+def _speakPlaceholderIfEmpty(info, obj, reason, index):
```
The `index` parameter is not required.

jcsteh commented on source/virtualBuffers/__init__.py:
```diff
@@ -218,6 +218,37 @@ def _getTextRange(self,start,end):
 
+	def _getPlaceholderAttribute(self, attrs, placeholderAttrsKey):
+		"""Gets the placeholder attribute to be used.
+				@returns The placeholder attribute when there a value is not present in the related field.
```
Needs to be a colon after `@returns`. Also, I'd suggest this:

```
@return: The placeholder attribute when there is no content within the ControlField.
```

derekriemer replied:
There -> the in the above line

jcsteh commented on source/virtualBuffers/__init__.py:
```diff
@@ -218,6 +218,37 @@ def _getTextRange(self,start,end):
+		placeholder = attrs.get(placeholderAttrsKey)
+		# For efficiency, only check if it is valid to return placeholder when we have a placeholder value to return.
+		if not placeholder:
+			return None;
```
nit: Extraneous semi-colon.

jcsteh commented on source/virtualBuffers/__init__.py:
```diff
@@ -218,6 +218,37 @@ def _getTextRange(self,start,end):
+		# For efficiency, only check if it is valid to return placeholder when we have a placeholder value to return.
+		if not placeholder:
+			return None;
+		# Get the start and end offsets for the value. This can be used to check if the field has a value.
```
Suggest use of term "content" instead of "value".

jcsteh commented on source/virtualBuffers/__init__.py:
```diff
@@ -218,6 +218,37 @@ def _getTextRange(self,start,end):
+				return placeholder
+			# Because fetching the value of the field could result in a large amount of text
+			# we only do it in order to check for space, or newline.
+			# We first compare the length by comparing the offselts, if the length is less than 2 (ie
```
typo: offselts -> offsets

jcsteh commented on source/virtualBuffers/gecko_ia2.py:
```diff
@@ -26,6 +26,9 @@ def _normalizeControlField(self,attrs):
 		if ariaCurrent != None:
 			attrs['current']= ariaCurrent
+		placeholder = self._getPlaceholderAttribute(attrs, "IAccessible2::attribute_placeholder")
+		if placeholder != None:
```
This would be better as `if placeholder` or, if you want to be super explicit, `if placeholder is not None`. I see `!= None` happens for `ariaCurrent` as well, so probably worth fixing that one too.

feerrenrut replied:
Yeah I think its worth differentiating from an empty string, which would indicate an authoring or browser error. Those errors will be harder to track down if we consider None and "" as equivalent.

jcsteh reviewed (changes requested):
Thanks. Looks pretty good to me overall!

feerrenrut committed:
Removed check for newline for virtual buffers get placeholder check.

We only need to test for space, not newline.
Tested in Firefox and Chrome in browse mode.

feerrenrut committed:
Make ariaProperties an attribute in edge

Tested edge in focus and browse mode.

feerrenrut committed:
Move comment

feerrenrut committed:
Fix issue in edge focus mode.

Moving by character was not limited to the content of the field, this
meant that the field boundary could be overrun.

The algorithm used could also cause a problem when the use entered
something like "\n Hello there" which would be considered empty.

Instead we use the object to create the text info, this should know the
correct range for the textinfo. Doing this I have confirmed that in Edge
in focus mode, using the test cases outlined on PR #7055 that an empty
field only ever has the value of '\n'.

feerrenrut committed:
Fix some small review actions

Comments, indentation, naming, unused param, comparison with None.

feerrenrut commented on source/NVDAObjects/UIA/edge.py:
```diff
@@ -409,6 +448,26 @@ def _get_isCurrent(self):
+		ti.move(textInfos.UNIT_CHARACTER, 1, endPoint="end")
+		text = ti.text
+		# Edge can report a range of values for empty fields:
+		if text == " " or text == "\n" or text =="\n " or text=="\n\n" or text ==" \n":
```
This actually required a bit of rework, I was curious about the performance optimisation that checking `len` might have, but found that it actually takes longer than just checking `text in`. Not that it matters now the function has been re-written.

jcsteh commented on source/NVDAObjects/UIA/edge.py:
```diff
@@ -449,22 +452,26 @@ def _get_isCurrent(self):
+			# it is collapsed therefore it is empty.
+			# exit early so we do not have to do not have to fetch `ti.text` which
+			# is potentially costly to performance.
+			log.info("Reef ti is collapsed")
```
Remove this please. :)

jcsteh commented on source/NVDAObjects/UIA/edge.py:
```diff
@@ -449,22 +452,26 @@ def _get_isCurrent(self):
-		if text == " " or text == "\n" or text =="\n " or text=="\n\n" or text ==" \n":
+		# Also we can not move (repeatedly by one character) since this can overrun the end of the field in edge.
+		# So instead, we use self to make a text info (which should have the right range) and then use the edge
+		# specific _rangeObj.getText function to get a subset of the full range of characters.
```
I'd suggest changing this from "Edge specific" to "UIA specific", since that method is a UIA method.

jcsteh reviewed (approved):
Looks good!

feerrenrut committed:
review actions

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into i7004-addSupportForAriaPlaceholder

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into i7004-addSupportForAriaPlaceholder

Resolved Conflicts:
	source/braille.py
	source/speech.py

feerrenrut committed:
Remove placeholder from allowedProperties dict

The allowedProperties dictionary had placeholder=false as the
default. This is uncessary since when we get it False is the default
value if the key is not present.
